### ISSUE TITLE
Some test failures in spyder/plugins/editor/widgets/tests/test_warnings.py for Python 3.10

### ISSUE THREAD

**Issue by juliangilbey** [spyder-ide/spyder#17101]
### Issue Report Checklist

* [x] Searched the [issues page](https://github.com/spyder-ide/spyder/issues?q=is%3Aissue) for similar reports
* [x] Read the relevant sections of the [Spyder Troubleshooting Guide](https://github.com/spyder-ide/spyder/wiki/Troubleshooting-Guide-and-FAQ) and followed its advice
* [x] Reproduced the issue after updating with ``conda update spyder`` (or ``pip``, if not using Anaconda)
* [ ] Could not reproduce inside ``jupyter qtconsole`` (if console-related)
* [x] Tried basic troubleshooting (if a bug/error)
    * [ ] Restarted Spyder
    * [ ] Reset preferences with ``spyder --reset``
    * [ ] Reinstalled the latest version of [Anaconda](https://www.anaconda.com/download/)
    * [ ] Tried the other applicable steps from the Troubleshooting Guide
* [x] Completed the **Problem Description**, **Steps to Reproduce** and **Version** sections below

## Problem Description

There are 12 more failing tests with `--run-slow`, presumably many are to do with using Qt 5.15.2.  9 of them are in `spyder/plugins/editor/widgets/tests/test_warnings.py`; I am wondering whether the problem is similar to that in #17100, as it seems like some code linter is just not being correctly run.

### What steps reproduce the problem?

I have `pylint`, `pydocstyle` and `pycodestyle` installed.

```
(sid-minimal)jdg@euler:/build/spyder-jdg/build-area/spyder-5.2.1+dfsg1$ debian/tests/pytest --run-slow
Testing with python3.10:
Temporary home directory: /tmp/tmp.wtGEzveBRq
Pytest Arguments: ['-vv', '-rw', '--durations=10', '-W ignore::UserWarning', '--run-slow', '-k', 'not test_get_pydoc and not test_qtbug35861 and not test_run_python_script_in_terminal and not test_run_python_script_in_terminal_blank_wdir and not test_run_python_script_in_terminal_with_wdir_empty and not test_find_external_plugins and not test_load_time and not test_objectexplorer_collection_types and not test_pylint.py and not test_object_arrays_display and not test_profiler_config_dialog.py and not test_handle_exception']
============================= test session starts ==============================
platform linux -- Python 3.10.1, pytest-6.2.5, py-1.10.0, pluggy-0.13.0 -- /usr/bin/python3.10
cachedir: .pytest_cache
PyQt5 5.15.6 -- Qt runtime 5.15.2 -- Qt compiled 5.15.2
rootdir: /build/spyder-jdg/build-area/spyder-5.2.1+dfsg1, configfile: pytest.ini
plugins: flaky-3.7.0, xvfb-2.0.0, mock-3.6.1, cov-3.0.0, order-1.0.0, qt-4.0.2, lazy-fixture-0.6.3
collected 1191 items / 39 deselected / 2 skipped / 1150 selected               

spyder/app/tests/test_cli_options.py::test_get_options SKIPPED (Don'...) [  0%]
[... snip ...]
spyder/plugins/editor/widgets/tests/test_hints_and_calltips.py::test_get_hints_not_triggered["a string"] PASSED [  5%]
spyder/plugins/editor/widgets/tests/test_warnings.py::test_ignore_warnings FAILED [  5%]
spyder/plugins/editor/widgets/tests/test_warnings.py::test_adding_warnings PASSED [  5%]
spyder/plugins/editor/widgets/tests/test_warnings.py::test_move_warnings PASSED [  5%]
spyder/plugins/editor/widgets/tests/test_warnings.py::test_get_warnings FAILED [  5%]
spyder/plugins/editor/widgets/tests/test_warnings.py::test_update_warnings_after_delete_line FAILED [  5%]
spyder/plugins/editor/widgets/tests/test_warnings.py::test_update_warnings_after_closequotes FAILED [  5%]
spyder/plugins/editor/widgets/tests/test_warnings.py::test_update_warnings_after_closebrackets FAILED [  6%]
spyder/plugins/editor/widgets/tests/test_warnings.py::test_ignore_warnings_with_comments[#noqa] FAILED [  6%]
spyder/plugins/editor/widgets/tests/test_warnings.py::test_ignore_warnings_with_comments[# NOQA] FAILED [  6%]
spyder/plugins/editor/widgets/tests/test_warnings.py::test_ignore_warnings_with_comments[# analysis:ignore] FAILED [  6%]
spyder/plugins/editor/widgets/tests/test_warnings.py::test_ignore_warnings_with_comments[# no-work] FAILED [  6%]
spyder/plugins/completion/providers/languageserver/tests/test_client.py::test_didOpen[lsp_provider] SKIPPED [  6%]
[... snip ...]

=================================== FAILURES ===================================
_____________________________ test_ignore_warnings _____________________________

qtbot = <pytestqt.qtbot.QtBot object at 0x7fd328360730>
completions_codeeditor_linting = (<spyder.plugins.editor.widgets.codeeditor.CodeEditor object at 0x7fd39e33e200>, <spyder.plugins.completion.plugin.CompletionPlugin object at 0x7fd39dc7c5e0>)

    @pytest.mark.slow
    @pytest.mark.order(2)
    @flaky(max_runs=5)
    def test_ignore_warnings(qtbot, completions_codeeditor_linting):
        """Test that the editor is ignoring some warnings."""
        editor, completion_plugin = completions_codeeditor_linting
    
        # Set text in editor
        editor.set_text(TEXT)
    
        CONF.set('completions',
                 ('provider_configuration', 'lsp', 'values', 'pydocstyle/ignore'),
                 'D100')
    
        CONF.set('completions',
                 ('provider_configuration', 'lsp', 'values','pycodestyle/ignore'),
                 'E261')
    
        # After this call the manager needs to be reinitialized
        completion_plugin.after_configuration_update([])
        qtbot.wait(2000)
    
        # Notify changes
        with qtbot.waitSignal(editor.completions_response_signal, timeout=30000):
            editor.document_did_change()
    
        # Get current warnings
        qtbot.wait(2000)
        warnings = editor.get_current_warnings()
    
        expected = [['D103: Missing docstring in public function', 1],
                    ['W293 blank line contains whitespace', 2],
                    ["undefined name 's'", 5],
                    ["undefined name 'undefined_function'", 7],
                    ["W292 no newline at end of file", 7],
                    ["""E305 expected 2 blank lines after class or """
                     """function definition, found 0""", 7]]
    
        CONF.set('completions',
                 ('provider_configuration', 'lsp', 'values', 'pydocstyle/ignore'),
                 '')
    
        CONF.set('completions',
                 ('provider_configuration', 'lsp', 'values','pycodestyle/ignore'),
                 '')
    
        completion_plugin.after_configuration_update([])
        qtbot.wait(2000)
    
>       assert warnings == expected
E       assert [['D103: Missing docstring in public function', 1],\n ['W293 blank line contains whitespace', 2],\n ['W292 no newline at end of file', 7],\n ['E305 expected 2 blank lines after class or function definition, found 0', 7]] == [['D103: Missing docstring in public function', 1],\n ['W293 blank line contains whitespace', 2],\n ["undefined name 's'", 5],\n ["undefined name 'undefined_function'", 7],\n ['W292 no newline at end of file', 7],\n ['E305 expected 2 blank lines after class or function definition, found 0', 7]]
E         At index 2 diff: ['W292 no newline at end of file', 7] != ["undefined name 's'", 5]
E         Right contains 2 more items, first extra item: ['W292 no newline at end of file', 7]
E         Full diff:
E           [
E            ['D103: Missing docstring in public function',
E             1],
E            ['W293 blank line contains whitespace',
E             2],
E         -  ["undefined name 's'",
E         -   5],
E         -  ["undefined name 'undefined_function'",
E         -   7],
E            ['W292 no newline at end of file',
E             7],
E            ['E305 expected 2 blank lines after class or function definition, found 0',
E             7],
E           ]

/build/spyder-jdg/build-area/spyder-5.2.1+dfsg1/spyder/plugins/editor/widgets/tests/test_warnings.py:113: AssertionError
______________________________ test_get_warnings _______________________________

qtbot = <pytestqt.qtbot.QtBot object at 0x7fd328390d00>
completions_codeeditor_linting = (<spyder.plugins.editor.widgets.codeeditor.CodeEditor object at 0x7fd34c1c6950>, <spyder.plugins.completion.plugin.CompletionPlugin object at 0x7fd39dc7c5e0>)

    @pytest.mark.slow
    @pytest.mark.order(2)
    @flaky(max_runs=5)
    def test_get_warnings(qtbot, completions_codeeditor_linting):
        """Test that the editor is returning the right list of warnings."""
        editor, _ = completions_codeeditor_linting
    
        # Set text in editor
        editor.set_text(TEXT)
    
        # Notify changes
        with qtbot.waitSignal(editor.completions_response_signal, timeout=30000):
            editor.document_did_change()
    
        # Wait for linting info to arrive
        qtbot.wait(2000)
    
        # Get current warnings
        warnings = editor.get_current_warnings()
    
        expected = [['D100: Missing docstring in public module', 1],
                    ['D103: Missing docstring in public function', 1],
                    ['W293 blank line contains whitespace', 2],
                    ['E261 at least two spaces before inline comment', 3],
                    ["undefined name 's'", 5],
                    ["undefined name 'undefined_function'", 7],
                    ["W292 no newline at end of file", 7],
                    ["""E305 expected 2 blank lines after class or """
                     """function definition, found 0""", 7]]
    
>       assert warnings == expected
E       assert [['D100: Missing docstring in public module', 1],\n ['D103: Missing docstring in public function', 1],\n ['W293 blank line contains whitespace', 2],\n ['E261 at least two spaces before inline comment', 3],\n ['W292 no newline at end of file', 7],\n ['E305 expected 2 blank lines after class or function definition, found 0', 7]] == [['D100: Missing docstring in public module', 1],\n ['D103: Missing docstring in public function', 1],\n ['W293 blank line contains whitespace', 2],\n ['E261 at least two spaces before inline comment', 3],\n ["undefined name 's'", 5],\n ["undefined name 'undefined_function'", 7],\n ['W292 no newline at end of file', 7],\n ['E305 expected 2 blank lines after class or function definition, found 0', 7]]
E         At index 4 diff: ['W292 no newline at end of file', 7] != ["undefined name 's'", 5]
E         Right contains 2 more items, first extra item: ['W292 no newline at end of file', 7]
E         Full diff:
E           [
E            ['D100: Missing docstring in public module',
E             1],
E            ['D103: Missing docstring in public function',
E             1],
E            ['W293 blank line contains whitespace',
E             2],
E            ['E261 at least two spaces before inline comment',
E             3],
E         -  ["undefined name 's'",
E         -   5],
E         -  ["undefined name 'undefined_function'",
E         -   7],
E            ['W292 no newline at end of file',
E             7],
E            ['E305 expected 2 blank lines after class or function definition, found 0',
E             7],
E           ]

/build/spyder-jdg/build-area/spyder-5.2.1+dfsg1/spyder/plugins/editor/widgets/tests/test_warnings.py:217: AssertionError
____________________ test_update_warnings_after_delete_line ____________________

qtbot = <pytestqt.qtbot.QtBot object at 0x7fd34f7537c0>
completions_codeeditor_linting = (<spyder.plugins.editor.widgets.codeeditor.CodeEditor object at 0x7fd39dc41990>, <spyder.plugins.completion.plugin.CompletionPlugin object at 0x7fd39dc7c5e0>)

    @pytest.mark.slow
    @pytest.mark.order(2)
    @flaky(max_runs=5)
    def test_update_warnings_after_delete_line(qtbot, completions_codeeditor_linting):
        """
        Test that code style warnings are correctly updated after deleting a line
        in the Editor.
    
        Regression test for spyder-ide/spyder#9299.
        """
        editor, _ = completions_codeeditor_linting
        editor.set_text(TEXT)
    
        # Notify changes.
        with qtbot.waitSignal(editor.completions_response_signal, timeout=30000):
            editor.document_did_change()
    
        # Wait for linting info to arrive
        qtbot.wait(2000)
    
        # Delete the blank line that is causing the W293 warning on line 2.
        editor.go_to_line(2)
        editor.delete_line()
    
        # Wait for the lsp_response_signal.
        qtbot.waitSignal(editor.completions_response_signal, timeout=30000)
    
        # Assert that the W293 warning is gone.
        expected = [['D100: Missing docstring in public module', 1],
                    ['D103: Missing docstring in public function', 1],
                    ['E261 at least two spaces before inline comment', 2],
                    ["undefined name 's'", 4],
                    ["undefined name 'undefined_function'", 6],
                    ["W292 no newline at end of file", 6],
                    ["""E305 expected 2 blank lines after class or """
                     """function definition, found 0""", 6]]
>       assert editor.get_current_warnings() == expected
E       assert [['D100: Missing docstring in public module', 1],\n ['D103: Missing docstring in public function', 1],\n ['E261 at least two spaces before inline comment', 2],\n ['W292 no newline at end of file', 6],\n ['E305 expected 2 blank lines after class or function definition, found 0', 6]] == [['D100: Missing docstring in public module', 1],\n ['D103: Missing docstring in public function', 1],\n ['E261 at least two spaces before inline comment', 2],\n ["undefined name 's'", 4],\n ["undefined name 'undefined_function'", 6],\n ['W292 no newline at end of file', 6],\n ['E305 expected 2 blank lines after class or function definition, found 0', 6]]
E         At index 3 diff: ['W292 no newline at end of file', 6] != ["undefined name 's'", 4]
E         Right contains 2 more items, first extra item: ['W292 no newline at end of file', 6]
E         Full diff:
E           [
E            ['D100: Missing docstring in public module',
E             1],
E            ['D103: Missing docstring in public function',
E             1],
E            ['E261 at least two spaces before inline comment',
E             2],
E         -  ["undefined name 's'",
E         -   4],
E         -  ["undefined name 'undefined_function'",
E         -   6],
E            ['W292 no newline at end of file',
E             6],
E            ['E305 expected 2 blank lines after class or function definition, found 0',
E             6],
E           ]

/build/spyder-jdg/build-area/spyder-5.2.1+dfsg1/spyder/plugins/editor/widgets/tests/test_warnings.py:256: AssertionError
____________________ test_update_warnings_after_closequotes ____________________

qtbot = <pytestqt.qtbot.QtBot object at 0x7fd388627940>
completions_codeeditor_linting = (<spyder.plugins.editor.widgets.codeeditor.CodeEditor object at 0x7fd34c1e97e0>, <spyder.plugins.completion.plugin.CompletionPlugin object at 0x7fd39dc7c5e0>)

    @pytest.mark.slow
    @pytest.mark.order(2)
    @flaky(max_runs=5)
    def test_update_warnings_after_closequotes(qtbot, completions_codeeditor_linting):
        """
        Test that code errors are correctly updated after activating closequotes
        in the Editor.
    
        Regression test for spyder-ide/spyder#9323.
        """
        editor, _ = completions_codeeditor_linting
        editor.textCursor().insertText("print('test)\n")
    
        expected = [['EOL while scanning string literal', 1]]
    
        # Notify changes.
        with qtbot.waitSignal(editor.completions_response_signal, timeout=30000):
            editor.document_did_change()
    
        # Wait for linting info to arrive
        qtbot.wait(2000)
    
>       assert editor.get_current_warnings() == expected
E       AssertionError: assert [] == [['EOL while scanning string literal', 1]]
E         Right contains one more item: ['EOL while scanning string literal', 1]
E         Full diff:
E         - [['EOL while scanning string literal', 1]]
E         + []

/build/spyder-jdg/build-area/spyder-5.2.1+dfsg1/spyder/plugins/editor/widgets/tests/test_warnings.py:281: AssertionError
___________________ test_update_warnings_after_closebrackets ___________________

qtbot = <pytestqt.qtbot.QtBot object at 0x7fd34c155150>
completions_codeeditor_linting = (<spyder.plugins.editor.widgets.codeeditor.CodeEditor object at 0x7fd3680700d0>, <spyder.plugins.completion.plugin.CompletionPlugin object at 0x7fd39dc7c5e0>)

    @pytest.mark.slow
    @pytest.mark.order(2)
    @flaky(max_runs=5)
    def test_update_warnings_after_closebrackets(qtbot, completions_codeeditor_linting):
        """
        Test that code errors are correctly updated after activating closebrackets
        in the Editor.
    
        Regression test for spyder-ide/spyder#9323.
        """
        editor, _ = completions_codeeditor_linting
        editor.textCursor().insertText("print('test'\n")
    
        expected = [['unexpected EOF while parsing', 1],
                    ['E901 TokenError: EOF in multi-line statement', 2]]
    
        # Notify changes.
        with qtbot.waitSignal(editor.completions_response_signal, timeout=30000):
            editor.document_did_change()
    
        # Wait for linting info to arrive
        qtbot.wait(2000)
    
>       assert editor.get_current_warnings() == expected
E       AssertionError: assert [['E901 TokenError: EOF in multi-line statement', 2]] == [['unexpected EOF while parsing', 1],\n ['E901 TokenError: EOF in multi-line statement', 2]]
E         At index 0 diff: ['E901 TokenError: EOF in multi-line statement', 2] != ['unexpected EOF while parsing', 1]
E         Right contains one more item: ['E901 TokenError: EOF in multi-line statement', 2]
E         Full diff:
E           [
E         -  ['unexpected EOF while parsing',
E         -   1],
E            ['E901 TokenError: EOF in multi-line statement',
E             2],
E           ]

/build/spyder-jdg/build-area/spyder-5.2.1+dfsg1/spyder/plugins/editor/widgets/tests/test_warnings.py:319: AssertionError
__________________ test_ignore_warnings_with_comments[#noqa] ___________________

qtbot = <pytestqt.qtbot.QtBot object at 0x7fd34c1f4e80>
completions_codeeditor_linting = (<spyder.plugins.editor.widgets.codeeditor.CodeEditor object at 0x7fd34c204a60>, <spyder.plugins.completion.plugin.CompletionPlugin object at 0x7fd39dc7c5e0>)
ignore_comment = '#noqa'

    @pytest.mark.slow
    @pytest.mark.order(2)
    @flaky(max_runs=5)
    @pytest.mark.parametrize(
        'ignore_comment', ['#noqa', '# NOQA', '# analysis:ignore', '# no-work']
    )
    def test_ignore_warnings_with_comments(
        qtbot, completions_codeeditor_linting, ignore_comment
    ):
        """
        Test that ignoring linting messages with inline comments works as
        expected.
        """
        editor, _ = completions_codeeditor_linting
        editor.textCursor().insertText(
            f"foo {ignore_comment}\n"
            "bar\n"
        )
    
        if ignore_comment == '# no-work':
            expected = [
                ["undefined name 'foo'", 1],
                ['D100: Missing docstring in public module', 1],
                ['E261 at least two spaces before inline comment', 1],
                ["undefined name 'bar'", 2]
            ]
        else:
            expected = [["undefined name 'bar'", 2]]
    
        # Notify changes.
        with qtbot.waitSignal(editor.completions_response_signal, timeout=30000):
            editor.document_did_change()
    
        # Wait for linting info to arrive
        qtbot.wait(2000)
    
>       assert editor.get_current_warnings() == expected
E       assert [] == [["undefined name 'bar'", 2]]
E         Right contains one more item: ["undefined name 'bar'", 2]
E         Full diff:
E         - [["undefined name 'bar'", 2]]
E         + []

/build/spyder-jdg/build-area/spyder-5.2.1+dfsg1/spyder/plugins/editor/widgets/tests/test_warnings.py:370: AssertionError
__________________ test_ignore_warnings_with_comments[# NOQA] __________________

qtbot = <pytestqt.qtbot.QtBot object at 0x7fd39dce1180>
completions_codeeditor_linting = (<spyder.plugins.editor.widgets.codeeditor.CodeEditor object at 0x7fd328343b50>, <spyder.plugins.completion.plugin.CompletionPlugin object at 0x7fd39dc7c5e0>)
ignore_comment = '# NOQA'

    @pytest.mark.slow
    @pytest.mark.order(2)
    @flaky(max_runs=5)
    @pytest.mark.parametrize(
        'ignore_comment', ['#noqa', '# NOQA', '# analysis:ignore', '# no-work']
    )
    def test_ignore_warnings_with_comments(
        qtbot, completions_codeeditor_linting, ignore_comment
    ):
        """
        Test that ignoring linting messages with inline comments works as
        expected.
        """
        editor, _ = completions_codeeditor_linting
        editor.textCursor().insertText(
            f"foo {ignore_comment}\n"
            "bar\n"
        )
    
        if ignore_comment == '# no-work':
            expected = [
                ["undefined name 'foo'", 1],
                ['D100: Missing docstring in public module', 1],
                ['E261 at least two spaces before inline comment', 1],
                ["undefined name 'bar'", 2]
            ]
        else:
            expected = [["undefined name 'bar'", 2]]
    
        # Notify changes.
        with qtbot.waitSignal(editor.completions_response_signal, timeout=30000):
            editor.document_did_change()
    
        # Wait for linting info to arrive
        qtbot.wait(2000)
    
>       assert editor.get_current_warnings() == expected
E       assert [] == [["undefined name 'bar'", 2]]
E         Right contains one more item: ["undefined name 'bar'", 2]
E         Full diff:
E         - [["undefined name 'bar'", 2]]
E         + []

/build/spyder-jdg/build-area/spyder-5.2.1+dfsg1/spyder/plugins/editor/widgets/tests/test_warnings.py:370: AssertionError
____________ test_ignore_warnings_with_comments[# analysis:ignore] _____________

qtbot = <pytestqt.qtbot.QtBot object at 0x7fd39d28abc0>
completions_codeeditor_linting = (<spyder.plugins.editor.widgets.codeeditor.CodeEditor object at 0x7fd39d29b010>, <spyder.plugins.completion.plugin.CompletionPlugin object at 0x7fd39dc7c5e0>)
ignore_comment = '# analysis:ignore'

    @pytest.mark.slow
    @pytest.mark.order(2)
    @flaky(max_runs=5)
    @pytest.mark.parametrize(
        'ignore_comment', ['#noqa', '# NOQA', '# analysis:ignore', '# no-work']
    )
    def test_ignore_warnings_with_comments(
        qtbot, completions_codeeditor_linting, ignore_comment
    ):
        """
        Test that ignoring linting messages with inline comments works as
        expected.
        """
        editor, _ = completions_codeeditor_linting
        editor.textCursor().insertText(
            f"foo {ignore_comment}\n"
            "bar\n"
        )
    
        if ignore_comment == '# no-work':
            expected = [
                ["undefined name 'foo'", 1],
                ['D100: Missing docstring in public module', 1],
                ['E261 at least two spaces before inline comment', 1],
                ["undefined name 'bar'", 2]
            ]
        else:
            expected = [["undefined name 'bar'", 2]]
    
        # Notify changes.
        with qtbot.waitSignal(editor.completions_response_signal, timeout=30000):
            editor.document_did_change()
    
        # Wait for linting info to arrive
        qtbot.wait(2000)
    
>       assert editor.get_current_warnings() == expected
E       assert [] == [["undefined name 'bar'", 2]]
E         Right contains one more item: ["undefined name 'bar'", 2]
E         Full diff:
E         - [["undefined name 'bar'", 2]]
E         + []

/build/spyder-jdg/build-area/spyder-5.2.1+dfsg1/spyder/plugins/editor/widgets/tests/test_warnings.py:370: AssertionError
________________ test_ignore_warnings_with_comments[# no-work] _________________

qtbot = <pytestqt.qtbot.QtBot object at 0x7fd34f770c10>
completions_codeeditor_linting = (<spyder.plugins.editor.widgets.codeeditor.CodeEditor object at 0x7fd388493520>, <spyder.plugins.completion.plugin.CompletionPlugin object at 0x7fd39d2af010>)
ignore_comment = '# no-work'

    @pytest.mark.slow
    @pytest.mark.order(2)
    @flaky(max_runs=5)
    @pytest.mark.parametrize(
        'ignore_comment', ['#noqa', '# NOQA', '# analysis:ignore', '# no-work']
    )
    def test_ignore_warnings_with_comments(
        qtbot, completions_codeeditor_linting, ignore_comment
    ):
        """
        Test that ignoring linting messages with inline comments works as
        expected.
        """
        editor, _ = completions_codeeditor_linting
        editor.textCursor().insertText(
            f"foo {ignore_comment}\n"
            "bar\n"
        )
    
        if ignore_comment == '# no-work':
            expected = [
                ["undefined name 'foo'", 1],
                ['D100: Missing docstring in public module', 1],
                ['E261 at least two spaces before inline comment', 1],
                ["undefined name 'bar'", 2]
            ]
        else:
            expected = [["undefined name 'bar'", 2]]
    
        # Notify changes.
        with qtbot.waitSignal(editor.completions_response_signal, timeout=30000):
            editor.document_did_change()
    
        # Wait for linting info to arrive
        qtbot.wait(2000)
    
>       assert editor.get_current_warnings() == expected
E       assert [['D100: Missing docstring in public module', 1],\n ['E261 at least two spaces before inline comment', 1]] == [["undefined name 'foo'", 1],\n ['D100: Missing docstring in public module', 1],\n ['E261 at least two spaces before inline comment', 1],\n ["undefined name 'bar'", 2]]
E         At index 0 diff: ['D100: Missing docstring in public module', 1] != ["undefined name 'foo'", 1]
E         Right contains 2 more items, first extra item: ['E261 at least two spaces before inline comment', 1]
E         Full diff:
E           [
E         -  ["undefined name 'foo'",
E         -   1],
E            ['D100: Missing docstring in public module',
E             1],
E            ['E261 at least two spaces before inline comment',
E             1],
E         -  ["undefined name 'bar'",
E         -   2],
E           ]

/build/spyder-jdg/build-area/spyder-5.2.1+dfsg1/spyder/plugins/editor/widgets/tests/test_warnings.py:370: AssertionError
[... snip ...]
```

## Versions

* Spyder version: 5.15.2
* Python version: 3.9 and 3.10
* Qt version: 5.15.2
* PyQt version: 5.15.6+dfsg-1+b1
* Operating System name/version: Debian testing and unstable

### Dependencies

```
# Mandatory:
atomicwrites >=1.2.0          :  1.4.0 (OK)
chardet >=2.0.0               :  4.0.0 (OK)
cloudpickle >=0.5.0           :  2.0.0 (OK)
cookiecutter >=1.6.0          :  1.7.3 (OK)
diff_match_patch >=20181111   :  20200713 (OK)
intervaltree >=3.0.2          :  3.0.2 (OK)
IPython >=7.6.0               :  7.27.0 (OK)
jedi >=0.17.2;<0.19.0         :  0.18.0 (OK)
jellyfish >=0.7               :  0.8.9 (OK)
jsonschema >=3.2.0            :  3.2.0 (OK)
keyring >=17.0.0              :  23.3.0 (OK)
nbconvert >=4.0               :  6.1.0 (OK)
numpydoc >=0.6.0              :  1.1.0 (OK)
parso >=0.7.0;<0.9.0          :  0.8.1 (OK)
pexpect >=4.4.0               :  4.8.0 (OK)
pickleshare >=0.4             :  0.7.5 (OK)
psutil >=5.3                  :  5.8.0 (OK)
pygments >=2.0                :  2.7.1 (OK)
pylint >=2.5.0                :  2.12.2 (OK)
pyls_spyder >=0.4.0           :  0.4.0 (OK)
pylsp >=1.3.2;<1.4.0          :  1.3.3 (OK)
pylsp_black >=1.0.0           :  1.0.1 (OK)
qdarkstyle =3.0.2             :  3.0.2 (OK)
qstylizer >=0.1.10            :  0.2.1 (OK)
qtawesome >=1.0.2             :  1.1.1 (OK)
qtconsole >=5.2.1;<5.3.0      :  5.2.1 (OK)
qtpy >=1.5.0                  :  1.9.0 (OK)
rtree >=0.9.7                 :  0.9.7 (OK)
setuptools >=49.6.0           :  58.2.0 (OK)
sphinx >=0.6.6                :  4.3.2 (OK)
spyder_kernels >=2.2.0;<2.3.0 :  2.2.0 (OK)
textdistance >=4.2.0          :  4.2.2 (OK)
three_merge >=0.1.1           :  0.1.1 (OK)
watchdog >=0.10.3             :  2.1.6 (OK)
xdg >=0.26                    :  0.27 (OK)
zmq >=17                      :  22.3.0 (OK)

# Optional:
cython >=0.21                 :  0.29.24 (OK)
matplotlib >=2.0.0            :  3.3.4 (OK)
numpy >=1.7                   :  1.21.5 (OK)
pandas >=1.1.1                :  1.3.4 (OK)
scipy >=0.17.0                :  1.7.1 (OK)
sympy >=0.7.3                 :  1.7.1 (OK)
```

**Comment by ccordoba12** [comment]
I found some errors when running these tests in Python 3.10, but they are not the same as yours. Instead, they are related to different linting messages, probably introduced in newer Python versions.

This part in your logs

```
E         -  ["undefined name 's'",
E         -   5],
E         -  ["undefined name 'undefined_function'",
E         -   7],
```

indicates to me installation issues in your side (probably with Pyflakes). So please review that.

**Comment by juliangilbey** [comment]
Ah, thanks @ccordoba12 - that solves most of the issues.  Pyflakes is not mentioned anywhere in the list of requirements or test requirements (though it is mentioned in the documentation), so I had not installed it - I am testing in a minimal environment for the best chance of reproducibility.  (The same goes for `autopep8`, `pydocstyle`, `pycodestyle` and `yapf`, by the way: at least `autopep8` and `yapf` are used in some of the tests, and `pydocstyle` and `pycodestyle` might be assumed, though I'm less certain about those.)

Now only two tests fail, and it does seem to be about differently worded warning messages in Python 3.10.  Here they are (minus the boring flaky warnings):
```
============================= test session starts ==============================
platform linux -- Python 3.10.1, pytest-6.2.5, py-1.10.0, pluggy-0.13.0 -- /usr/bin/python3.10
cachedir: .pytest_cache
PyQt5 5.15.6 -- Qt runtime 5.15.2 -- Qt compiled 5.15.2
rootdir: /build/spyder-jdg/build-area/spyder-5.2.1+dfsg1, configfile: pytest.ini
plugins: flaky-3.7.0, xvfb-2.0.0, mock-3.6.1, cov-3.0.0, order-1.0.0, qt-4.0.2, lazy-fixture-0.6.3
collected 11 items                                                             

spyder/plugins/editor/widgets/tests/test_warnings.py::test_ignore_warnings PASSED [  9%]
spyder/plugins/editor/widgets/tests/test_warnings.py::test_adding_warnings PASSED [ 18%]
spyder/plugins/editor/widgets/tests/test_warnings.py::test_move_warnings PASSED [ 27%]
spyder/plugins/editor/widgets/tests/test_warnings.py::test_get_warnings PASSED [ 36%]
spyder/plugins/editor/widgets/tests/test_warnings.py::test_update_warnings_after_delete_line PASSED [ 45%]
spyder/plugins/editor/widgets/tests/test_warnings.py::test_update_warnings_after_closequotes FAILED [ 54%]
spyder/plugins/editor/widgets/tests/test_warnings.py::test_update_warnings_after_closebrackets FAILED [ 63%]
spyder/plugins/editor/widgets/tests/test_warnings.py::test_ignore_warnings_with_comments[#noqa] PASSED [ 72%]
spyder/plugins/editor/widgets/tests/test_warnings.py::test_ignore_warnings_with_comments[# NOQA] PASSED [ 81%]
spyder/plugins/editor/widgets/tests/test_warnings.py::test_ignore_warnings_with_comments[# analysis:ignore] PASSED [ 90%]
spyder/plugins/editor/widgets/tests/test_warnings.py::test_ignore_warnings_with_comments[# no-work] PASSED [100%]

=================================== FAILURES ===================================
____________________ test_update_warnings_after_closequotes ____________________

qtbot = <pytestqt.qtbot.QtBot object at 0x7f1ce0406fe0>
completions_codeeditor_linting = (<spyder.plugins.editor.widgets.codeeditor.CodeEditor object at 0x7f1ce0148550>, <spyder.plugins.completion.plugin.CompletionPlugin object at 0x7f1d446b1fc0>)

    @pytest.mark.slow
    @pytest.mark.order(2)
    @flaky(max_runs=5)
    def test_update_warnings_after_closequotes(qtbot, completions_codeeditor_linting):
        """
        Test that code errors are correctly updated after activating closequotes
        in the Editor.
    
        Regression test for spyder-ide/spyder#9323.
        """
        editor, _ = completions_codeeditor_linting
        editor.textCursor().insertText("print('test)\n")
    
        expected = [['EOL while scanning string literal', 1]]
    
        # Notify changes.
        with qtbot.waitSignal(editor.completions_response_signal, timeout=30000):
            editor.document_did_change()
    
        # Wait for linting info to arrive
        qtbot.wait(2000)
    
>       assert editor.get_current_warnings() == expected
E       AssertionError: assert [['unterminated string literal (detected at line 1)', 1]] == [['EOL while scanning string literal', 1]]
E         At index 0 diff: ['unterminated string literal (detected at line 1)', 1] != ['EOL while scanning string literal', 1]
E         Full diff:
E         - [['EOL while scanning string literal', 1]]
E         + [['unterminated string literal (detected at line 1)', 1]]

spyder/plugins/editor/widgets/tests/test_warnings.py:281: AssertionError
___________________ test_update_warnings_after_closebrackets ___________________

qtbot = <pytestqt.qtbot.QtBot object at 0x7f1ce0108250>
completions_codeeditor_linting = (<spyder.plugins.editor.widgets.codeeditor.CodeEditor object at 0x7f1d0155e7a0>, <spyder.plugins.completion.plugin.CompletionPlugin object at 0x7f1d446b1fc0>)

    @pytest.mark.slow
    @pytest.mark.order(2)
    @flaky(max_runs=5)
    def test_update_warnings_after_closebrackets(qtbot, completions_codeeditor_linting):
        """
        Test that code errors are correctly updated after activating closebrackets
        in the Editor.
    
        Regression test for spyder-ide/spyder#9323.
        """
        editor, _ = completions_codeeditor_linting
        editor.textCursor().insertText("print('test'\n")
    
        expected = [['unexpected EOF while parsing', 1],
                    ['E901 TokenError: EOF in multi-line statement', 2]]
    
        # Notify changes.
        with qtbot.waitSignal(editor.completions_response_signal, timeout=30000):
            editor.document_did_change()
    
        # Wait for linting info to arrive
        qtbot.wait(2000)
    
>       assert editor.get_current_warnings() == expected
E       assert [["'(' was never closed", 1],\n ['E901 TokenError: EOF in multi-line statement', 2]] == [['unexpected EOF while parsing', 1],\n ['E901 TokenError: EOF in multi-line statement', 2]]
E         At index 0 diff: ["'(' was never closed", 1] != ['unexpected EOF while parsing', 1]
E         Full diff:
E           [
E         -  ['unexpected EOF while parsing', 1],
E         +  ["'(' was never closed", 1],
E            ['E901 TokenError: EOF in multi-line statement', 2],
E           ]

spyder/plugins/editor/widgets/tests/test_warnings.py:319: AssertionError
```

Best wishes!

**Comment by ccordoba12** [comment]
> Ah, thanks @ccordoba12 - that solves most of the issues. Pyflakes is not mentioned anywhere in the list of requirements or test requirements

That's a transitive dependency (note the `all` below):

    Spyder -> python-lsp-server[all] -> Pyflakes

> Now only two tests fail, and it does seem to be about differently worded warning messages in Python 3.10

Yeah, I fixed those already.

**Comment by juliangilbey** [comment]
Ah, I hadn't noticed the `all` in the dependencies; the Debian `python-lsp-server` package has not included those dependencies.  I'll fix that on our side now then.

**Comment by juliangilbey** [comment]
And if you've fixed those two Python 3.10 issues already, then this issue can be closed - thanks so much for your speedy response!!

**Comment by ccordoba12** [comment]
It'll be closed once I merge the corresponding PR (I haven't pushed yet).
